cmd/docker/docker.go

```diff
@@ -44,15 +37,32 @@ func newDockerCommand(dockerCli *command.DockerCli) *cli.TopLevelCommand {
 				return command.ShowHelp(dockerCli.Err())(cmd, args)
 			}
 			return fmt.Errorf("docker: '%s' is not a docker command.\nSee 'docker --help'", args[0])
-
 		},
 		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			// Handle shell completion for plugin commands by forwarding the
+			// arguments to the plugin. Supporting completion is currently
+			// optional for plugins, so we ignore errors (if any).
+			if len(args) > 0 && cmd.Name() == cobra.ShellCompRequestCmd || cmd.Name() == cobra.ShellCompNoDescRequestCmd {
+				if _, _, err := cmd.Root().Find(args); err != nil {
+					// No such command. Try if we have a plugin that has this command
+					if err := tryPluginRun(dockerCli, cmd, args[0]); err == nil {
+						// We're done. Exit here to prevent the regular command
+						// from being executed.
+						os.Exit(0)
+					}
+					// Didn't find a plugin for this command either. Continue as
+					// usual, and have the default __complete / __completeNoDesc
+					// command handle the completion.
+					return nil
+				}
+			}
 			return isSupported(cmd, dockerCli)
 		},
 		Version:               fmt.Sprintf("%s, build %s", version.Version, version.GitCommit),
 		DisableFlagsInUseLine: true,
 	}
-	opts, flags, helpCmd = cli.SetupRootCommand(cmd)
+	opts, flags, helpCmd := cli.SetupRootCommand(cmd)
+	flags.String("completion", "", "Print the shell completion script for the specified shell (bash, fish, powershell, or zsh) and quit")
```

Comment:
question: we mention `powershell` here but it's not in the completion map in `completion.go`. I guess it would fail then ? 🤔

Reply:
Ok my bad, `powershell` is fully generated by cobra 👍

Reply:
No, it falls back to the auto-generated Cobra completion